pkg/apiserver/rest/usecase/application.go

```diff
@@ -1415,3 +1493,220 @@ func genWebhookToken() string {
+			if errors.Is(err, datastore.ErrRecordNotExist) {
+				continue
+			}
+			log.Logger.Warnf("delete app %s comp %s failure %s", appPrimaryKey, compName, err.Error())
+		}
+	}
+
+	for _, comp := range targetComps {
+		// add or update new app's components from old app
+		if utils.StringsContain(readyToAdd, comp.Name) || utils.StringsContain(readyToUpdate, comp.Name) {
+			compModel := convertToModelComponent(appPrimaryKey, comp)
+			properties, err := model.NewJSONStruct(comp.Properties)
+			if err != nil {
+				return &apisv1.AppResetResponse{}, bcode.ErrInvalidProperties
+			}
+			compModel.Properties = properties
+			if err := c.ds.Add(ctx, &compModel); err != nil {
+				if errors.Is(err, datastore.ErrRecordExist) {
+					err := c.ds.Put(ctx, &compModel)
+					if err != nil {
+						log.Logger.Warnf("update comp %s  for app %s failure %s", comp.Name, utils2.Sanitize(appPrimaryKey), err.Error())
+					}
+					return &apisv1.AppResetResponse{IsReset: true}, err
+				}
+				log.Logger.Warnf("add comp %s  for app %s failure %s", comp.Name, utils2.Sanitize(appPrimaryKey), err.Error())
+				return &apisv1.AppResetResponse{}, err
+			}
+		}
+	}
+	return &apisv1.AppResetResponse{IsReset: true}, nil
+}
+
+func dryRunApplication(ctx context.Context, app *v1beta1.Application) (bytes.Buffer, error) {
+	c := common2.Args{
+		Schema: common2.Scheme,
+	}
+
+	var buff = bytes.Buffer{}
+
+	newClient, err := c.GetClient()
+	if err != nil {
+		return buff, err
+	}
+	var objs []oam.Object
+	pd, err := c.GetPackageDiscover()
+	if err != nil {
+		return buff, err
+	}
+
+	dm, err := discoverymapper.New(c.Config)
+	if err != nil {
+		return buff, err
+	}
+
+	dryRunOpt := dryrun.NewDryRunOption(newClient, dm, pd, objs)
+	comps, err := dryRunOpt.ExecuteDryRun(ctx, app)
+	if err != nil {
+		return buff, errors.New("generate OAM objects")
+	}
+
+	var components = make(map[string]*unstructured.Unstructured)
+	for _, comp := range comps {
+		components[comp.Name] = comp.StandardWorkload
+	}
+	buff.Write([]byte(fmt.Sprintf("---\n# Application(%s) \n---\n\n", app.Name)))
+	result, err := yaml.Marshal(app)
+	if err != nil {
+		return buff, errors.New("marshal app error")
+	}
+	buff.Write(result)
+	buff.Write([]byte("\n---\n"))
+	for _, c := range comps {
+		buff.Write([]byte(fmt.Sprintf("---\n# Application(%s) -- Component(%s) \n---\n\n", app.Name, c.Name)))
+		result, err := yaml.Marshal(components[c.Name])
+		if err != nil {
+			return buff, errors.New("marshal result for component " + c.Name + " object in yaml format")
+		}
+		buff.Write(result)
+		buff.Write([]byte("\n---\n"))
+		for _, t := range c.Traits {
+			result, err := yaml.Marshal(t)
+			if err != nil {
+				return buff, errors.New("marshal result for component " + c.Name + " object in yaml format")
+			}
+			buff.Write(result)
+			buff.Write([]byte("\n---\n"))
+		}
+		buff.Write([]byte("\n"))
+	}
+	return buff, nil
+}
+
+func ignoreSomeParams(o *v1beta1.Application) {
```

Comment:
it seems to be a common function, do we have something similar in `vela dry-run/live-diff` package? Or make it as a utility function

Reply:
yes,  core code comes from: pkg/plugin/cli/livediff.go:29